--- a/Source/WebKit/qt/Api/qwebsettings.h	2013-04-09 02:11:38.000000000 +0200
+++ b/Source/WebKit/qt/Api/qwebsettings.h	2013-05-16 13:22:35.530464274 +0200
@@ -82,7 +82,8 @@
         CSSGridLayoutEnabled,
         ScrollAnimatorEnabled,
         CaretBrowsingEnabled,
-        NotificationsEnabled
+        NotificationsEnabled,
+        WebSecurityEnabled
     };
     enum WebGraphic {
         MissingImageGraphic,

--- a/Source/WebKit/qt/Api/qwebsettings.cpp	2013-04-09 02:11:39.000000000 +0200
+++ b/Source/WebKit/qt/Api/qwebsettings.cpp	2013-05-16 13:27:38.954452972 +0200
@@ -279,6 +279,11 @@
         settings->setNeedsSiteSpecificQuirks(value);
 
         settings->setUsesPageCache(WebCore::pageCache()->capacity());
+
+        value = attributes.value(QWebSettings::WebSecurityEnabled,
+                                     global->attributes.value(QWebSettings::WebSecurityEnabled));
+        settings->setWebSecurityEnabled(value);
+
     } else {
         QList<QWebSettingsPrivate*> settings = *::allSettings();
         for (int i = 0; i < settings.count(); ++i)
@@ -547,6 +552,7 @@
     d->offlineStorageDefaultQuota = 5 * 1024 * 1024;
     d->defaultTextEncoding = QLatin1String("iso-8859-1");
     d->thirdPartyCookiePolicy = AlwaysAllowThirdPartyCookies;
+    d->attributes.insert(QWebSettings::WebSecurityEnabled, true);
 }

