--- a/src/platformsupport/eglconvenience/eglconvenience.pri	2013-05-30 06:13:14.270529593 +0200
+++ b/src/platformsupport/eglconvenience/eglconvenience.pri	2013-05-30 06:15:25.010532166 +0200
@@ -8,6 +8,8 @@
         $$PWD/qeglplatformcontext.cpp \
         $$PWD/qeglpbuffer.cpp
 
+    CONFIG += egl
+
     contains(QT_CONFIG,xlib) {
         HEADERS += \
             $$PWD/qxlibeglintegration_p.h
