--- a/mkspecs/common/posix/qplatformdefs.h
+++ b/mkspecs/common/posix/qplatformdefs.h
@@ -123,7 +123,11 @@
 #define QT_READ                 ::read
 #define QT_WRITE                ::write
 
+#ifdef QT_LARGEFILE_SUPPORT
 #define QT_OPEN_LARGEFILE       O_LARGEFILE
+#else
+#define QT_OPEN_LARGEFILE       0
+#endif
 #define QT_OPEN_RDONLY          O_RDONLY
 #define QT_OPEN_WRONLY          O_WRONLY
 #define QT_OPEN_RDWR            O_RDWR
