--- a/src/plugins/platforms/qwayland-brcm-egl/qwaylandbrcmeglwindow.h	2013-06-03 13:26:59.353707910 +0200
+++ b/src/plugins/platforms/qwayland-brcm-egl/qwaylandbrcmeglwindow.h	2013-06-03 13:27:52.525707699 +0200
@@ -54,7 +54,7 @@
 class QWaylandGLContext;
 class QWaylandBrcmBuffer;
 
-class QWaylandBrcmEglWindow : public QObject, public QWaylandWindow
+class QWaylandBrcmEglWindow : public QWaylandWindow
 {
     Q_OBJECT
 public:
